src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -144,20 +142,102 @@ public async void BeginInvokeDotNetFromJS(string callId, string assemblyName, st
             }
         }
 
-        public async ValueTask DisposeAsync()
+        private async Task OnCircuitOpenedAsync(CancellationToken cancellationToken)
         {
-            await Renderer.InvokeAsync(async () =>
+            for (var i = 0; i < _circuitHandlers.Length; i++)
             {
+                var circuitHandler = _circuitHandlers[i];
+                try
+                {
+                    await circuitHandler.OnCircuitOpenedAsync(Circuit, cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    OnHandlerError(circuitHandler, nameof(CircuitHandler.OnCircuitOpenedAsync), ex);
+                }
+            }
+        }
+
+        public async Task OnConnectionUpAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                await HandlerLock.WaitAsync(cancellationToken);
+
                 for (var i = 0; i < _circuitHandlers.Length; i++)
                 {
-                    await _circuitHandlers[i].OnConnectionDownAsync(Circuit, default);
+                    var circuitHandler = _circuitHandlers[i];
+                    try
+                    {
+                        await circuitHandler.OnConnectionUpAsync(Circuit, cancellationToken);
+                    }
+                    catch (Exception ex)
+                    {
+                        OnHandlerError(circuitHandler, nameof(CircuitHandler.OnConnectionUpAsync), ex);
+                    }
                 }
+            }
+            finally
+            {
+                HandlerLock.Release();
+            }
+        }
+
+        public async Task OnConnectionDownAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                await HandlerLock.WaitAsync(cancellationToken);
 
                 for (var i = 0; i < _circuitHandlers.Length; i++)
                 {
-                    await _circuitHandlers[i].OnCircuitClosedAsync(Circuit, default);
+                    var circuitHandler = _circuitHandlers[i];
+                    try
+                    {
+                        await circuitHandler.OnConnectionDownAsync(Circuit, cancellationToken);
+                    }
+                    catch (Exception ex)
+                    {
+                        OnHandlerError(circuitHandler, nameof(CircuitHandler.OnConnectionDownAsync), ex);
+                    }
                 }
-            });
+            }
+            finally
+            {
+                HandlerLock.Release();
+            }
+        }
+
+        protected virtual void OnHandlerError(CircuitHandler circuitHandler, string handlerMethod, Exception ex)
+        {
+            Log.UnhandledExceptionInvokingCircuitHandler(_logger, circuitHandler, handlerMethod, ex);
+        }
+
+        private async Task OnCircuitDownAsync()
+        {
+            for (var i = 0; i < _circuitHandlers.Length; i++)
```

Comment:
this isn't protected by the lock.. is this an issue? or is this protected by the design of the circuit pool?

Reply:
Neither of the circuit events have concurrency issues - the server controls when a circuit is being discarded so it doesn't need protection.